Incorrect linting with type parameters PEP695 (generic)

### Bug description

Python 3.12.0

Possibly sharing the same root cause as #9098 but I'm reporting as the behavior is different just in case (here no crashes, just incorrect results). It'd be great if pylint can support all the type parameters specification described in https://peps.python.org/pep-0695/.


```python
def constrained[T: (str, bytes)](argument: T) -> T:
    return argument
```

(EDIT: removed the second example in edit because it's not valid)

### Configuration

```ini
No pylintrc is used (under the default setting).
```


### Command used

```shell
pylint 695.py --disable=C0114,C0116

(Note: C0114, C0116 are missing docstrings)
```


### Pylint output

```python
************* Module 695
695.py:1:43: E0602: Undefined variable 'T' (undefined-variable)
695.py:1:49: E0602: Undefined variable 'T' (undefined-variable)
695.py:1:16: W0612: Unused variable 'T' (unused-variable)
```


### Expected behavior

There should be no E0602, W0612, E0602 warnings and errors. `T` in the parameter annotation is the generic type.

### Pylint version

```shell
pylint 3.0.0
astroid 3.0.0
Python 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:36:57) [Clang 15.0.7 ]
```


### OS / Environment

macOS and Linux

### Additional dependencies

```python
N/A
```

--- a/doc/whatsnew/fragments/9110.false_positive
+++ b/doc/whatsnew/fragments/9110.false_positive
@@ -1,0 +1,3 @@
+Fix ``used-before-assignment`` false positive for generic type syntax (PEP 695, Python 3.12).
+
+Closes #9110

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -120,6 +120,7 @@ NODES_WITH_VALUE_ATTR = (
     nodes.Expr,
     nodes.Return,
     nodes.Match,
+    nodes.TypeAlias,
 )
 
 
@@ -2328,6 +2329,8 @@ class VariablesChecker(BaseChecker):
         if isinstance(defstmt, nodes.Match):
             return any(case.guard for case in defstmt.cases)
         if isinstance(defstmt, nodes.IfExp):
+            return True
+        if isinstance(defstmt, nodes.TypeAlias):
             return True
         if isinstance(defstmt.value, nodes.BaseContainer):
             return any(

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -41,7 +41,7 @@ dependencies    = [
     # Also upgrade requirements_test_min.txt.
     # Pinned to dev of second minor update to allow editable installs and fix primer issues,
     # see https://github.com/pylint-dev/astroid/issues/1341
-    "astroid>=3.0.0,<=3.1.0-dev0",
+    "astroid>=3.0.1,<=3.1.0-dev0",
     "isort>=4.2.5,<6",
     "mccabe>=0.6,<0.8",
     "tomli>=1.1.0;python_version<'3.11'",

--- a/requirements_test_min.txt
+++ b/requirements_test_min.txt
@@ -1,6 +1,6 @@
 .[testutils,spelling]
 # astroid dependency is also defined in pyproject.toml
-astroid==3.0.0  # Pinned to a specific version for tests
+astroid==3.0.1  # Pinned to a specific version for tests
 typing-extensions~=4.8
 py~=1.11.0
 pytest~=7.4

--- a/tests/functional/u/used/used_before_assignment_py312.py
+++ b/tests/functional/u/used/used_before_assignment_py312.py
@@ -1,0 +1,6 @@
+"""used-before-assignment re: python 3.12 generic typing syntax (PEP 695)"""
+
+from typing import Callable
+type Point[T] = tuple[T, ...]
+type Alias[*Ts] = tuple[*Ts]
+type Alias[**P] = Callable[P]

--- a/tests/functional/u/used/used_before_assignment_py312.rc
+++ b/tests/functional/u/used/used_before_assignment_py312.rc
@@ -1,0 +1,2 @@
+[testoptions]
+min_pyver=3.12

